packages/shopping-cart/src/use-refetch-on-focus.ts

```diff
@@ -72,23 +83,27 @@ export default function useRefetchOnFocus(
 			}
 
 			debug( 'window was refocused; refetching' );
-			refetch();
+			reloadFromServer();
 		}
 
+		debug( 'adding focus listeners' );
 		window.addEventListener( 'visibilitychange', handleFocusChange );
 		window.addEventListener( 'focus', handleFocusChange );
 		window.addEventListener( 'online', handleFocusChange );
 
 		return () => {
+			debug( 'removing focus listeners' );
 			window.removeEventListener( 'visibilitychange', handleFocusChange );
 			window.removeEventListener( 'focus', handleFocusChange );
 			window.removeEventListener( 'online', handleFocusChange );
 		};
 	}, [
 		cartKey,
-		options.refetchOnWindowFocus,
+		refetchOnWindowFocus,
 		lastCart.cart_generated_at_timestamp,
-		refetch,
-		cacheStatus,
+		reloadFromServer,
+		isLoading,
+		isPendingUpdate,
+		loadingError,
```

Comment:
This was tricky because the listener needs to have all the data it requires in order to decide if a fetch is needed, but thanks to your feedback I realized that we only need to depend on basically one thing: the `ShoppingCartManager`, and that conveniently never changes (unless the cart key changes).

Reply:
You remove one important check, though: that the on-focus refetch is triggered only when another fetch is not already running, i.e., both `isLoading` (initial fetch) and `isPendingUpdate` (sync after change) are `false`.

That was done by uninstalling the listeners during the "is not valid" period and then installing them again. Somewhat inefficient, but it did the job.

Reply:
I handled that case in a different way: the reducer itself will now ignore reload requests if it's in a pending state, which it should really do anyway.